src/main/python/deep_qa/layers/attention/weighted_sum.py

```diff
@@ -0,0 +1,91 @@
+
+    While the above spec shows inputs with 3 and 2 modes, we also allow inputs of any order; we
+    always sum over the second-to-last dimension of the "matrix", weighted by the last dimension of
+    the "vector".  Higher-order tensors get complicated for matching things, though, so there is a
+    hard constraint: all dimensions in the "matrix" before the final embedding must be matched in
+    the "vector".
+
+    For example, say I have a "matrix" with dimensions (batch_size, num_queries, num_words,
+    embedding_dim), representing some kind of embedding or encoding of several multi-word queries.
+    My attention "vector" must then have at least those dimensions, and could have more.  So I
+    could have an attention over words per query, with shape (batch_size, num_queries, num_words),
+    or I could have an attention over query words for every document in some list, with shape
+    (batch_size, num_documents, num_queries, num_words).  Both of these cases are fine.  In the
+    first case, the returned tensor will have shape (batch_size, num_queries, embedding_dim), and
+    in the second case, it will have shape (batch_size, num_documents, num_queries, embedding_dim).
+    But you _can't_ have an attention "vector" that does not include all of the queries, so shape
+    (batch_size, num_words) is not allowed - you haven't specified how to handle that dimension in
+    the "matrix", so we can't do anything with this input.
+    '''
+    def __init__(self, **kwargs):
+        self.supports_masking = True
+        super(WeightedSum, self).__init__(**kwargs)
+
+    @overrides
+    def compute_mask(self, inputs, mask=None):
+        # pylint: disable=unused-argument
+        # We don't need to worry about a mask after we've summed over the rows of the matrix.
+        # You might actually still need a mask if you used a higher-order tensor, but probably the
+        # right place to handle that is with careful use of TimeDistributed.  Or submit a PR.
+        return None
+
+    @overrides
+    def get_output_shape_for(self, input_shapes):
+        matrix_shape, attention_shape = input_shapes
+        return attention_shape[:-1] + matrix_shape[-1:]
+
+    @overrides
+    def call(self, inputs, mask=None):
+        matrix, attention_vector = inputs
+        matrix = self._expand_matrix_if_necessary(matrix, attention_vector)
+        matrix_mask = mask[0]
+        if matrix_mask is not None:
+            matrix_mask = K.expand_dims(matrix_mask, dim=-1)
+            matrix_mask = self._expand_matrix_if_necessary(matrix_mask, attention_vector)
+            matrix = switch(matrix_mask, matrix, K.zeros_like(matrix))
+        return K.sum(K.expand_dims(attention_vector, dim=-1) * matrix, -2)
+
+    @staticmethod
+    def _expand_matrix_if_necessary(matrix, attention_vector):
+        try:
+            matrix_shape = K.int_shape(matrix)[:-1]  # taking off the embedding dimension here.
+        except Exception:  # pylint: disable=broad-except
```

Comment:
It's because this gets called on the mask, and the mask isn't made into a keras tensor.  Maybe we could fix this by doing whatever keras does to make things into keras tensors in the first place, and we could probably find out by tracing what happens to the output of `call()`.